src/server.rs

```diff
@@ -285,47 +350,172 @@ mod tests {
             ),
         };
 
-        let result = block_on(server.did_open(params));
+        block_on(server.did_open(params));
 
-        assert_eq!((), result);
+        assert_eq!(
+            vec![&lsp::Url::parse("file:///home/user/file.flux")
+                .unwrap()],
+            server
+                .store
+                .lock()
+                .unwrap()
+                .keys()
+                .collect::<Vec<&lsp::Url>>()
+        );
+        let uri =
+            lsp::Url::parse("file:///home/user/file.flux").unwrap();
+        let contents =
+            server.store.lock().unwrap().get(&uri).unwrap().clone();
+        assert_eq!("from(", contents);
+    }
+
+    #[test]
+    fn test_did_change() {
+        let server = create_server();
+        open_file(
+            &server,
+            r#"from(bucket: "bucket") |> first()"#.to_string(),
+        );
+
+        let params = lsp::DidChangeTextDocumentParams {
+            text_document: lsp::VersionedTextDocumentIdentifier {
+                uri: lsp::Url::parse("file:///home/user/file.flux")
+                    .unwrap(),
+                version: -2,
```

Comment:
I wonder if the version is intended to speed up communicating "undo" events where you might want to quickly revert to an earlier version of the document?

Reply:
It _claims_ that, kinda, but the server is never the authoritative version of the file anyway. In most of the other lsp servers I looked at, they either don't implement it, or have an issue where they keep adding versions that are never cleaned up, so your lsp could eventually OOM because the client was chatty with versions.

The best hypothesis I have for this is the functionality described as [TextDocumentEdit](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocumentEdit) in the spec, which is new in the most recent and current version of the spec, 3.16. It says

> Describes textual changes on a single text document. The text document is referred to as a OptionalVersionedTextDocumentIdentifier to allow clients to check the text document version before an edit is applied. A TextDocumentEdit describes all changes on a version Si and after they are applied move the document to version Si+1. So the creator of a TextDocumentEdit doesn’t need to sort the array of edits or do any kind of ordering. However the edits must be non overlapping.

That's the only place I've found where the client actually consumes the other side of the version, and we don't signal that support in either of our LSP servers right now.

In [VersionedTextDocumentIdentifier](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#versionedTextDocumentIdentifier), it hints at undo/redo, but also says the following about version:

> * The version number of a document will increase after each change,
> * including undo/redo. The number doesn't need to be consecutive.

If the version doesn't have to be consecutive, that would indicate to me that the data structure is more complex than what I've seen across other lsp implementations. I think it'd be neat to support it, but I'd rather keep us from peak complexity during a refactor.